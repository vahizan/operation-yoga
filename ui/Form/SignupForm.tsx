import { FormEvent, useState } from "react";
import { useRouter } from "next/navigation";
import axios from "axios";

import styles from "./signupForm.module.scss";
import BouncingDotsLoader from "../Loader/BouncingDotsLoader";

const SignupForm = () => {
  const [name, setName] = useState("");
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [phone, setPhone] = useState("");
  const [error, setError] = useState("");
  const [isSigningUp, setSigningUp] = useState<boolean>();

  const router = useRouter();

  const handleSubmit = async (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    setError("");

    if (!name || !email || !password) {
      setError("Please fill in all required fields.");
      setSigningUp(false);
      return;
    }

    try {
      setSigningUp(true);

      const response = await axios.post(
        "/api/auth/signup",
        JSON.stringify({ name, email, password, phone }),
        {
          headers: {
            "Content-Type": "application/json",
          },
        }
      );

      if (response.data) {
        // SignupForm successful, redirect to the login page
        await router.push("/login");
      } else {
        const data = await response.data.json();
        setError(data.message);
      }
    } catch (error) {
      console.log("error", error);
      setError("An error occurred. Please try again.");
    } finally {
      setSigningUp(false);
    }
  };

  return (
    <div className={styles.signupFormContainer}>
      <form onSubmit={handleSubmit}>
        <div className={styles.formGroup}>
          <label htmlFor={"name"}>Name:</label>
          <input
            id={"name"}
            name={"name"}
            type="text"
            value={name}
            onChange={(e) => setName(e.target.value)}
          />
        </div>

        <div className={styles.formGroup}>
          <label htmlFor={"email"}>Email:</label>
          <input
            name={"email"}
            id={"email"}
            type="email"
            value={email}
            onChange={(e) => setEmail(e.target.value)}
          />
        </div>

        <div className={styles.formGroup}>
          <label htmlFor={"password"}>Password:</label>
          <input
            name={"password"}
            id={"password"}
            type="password"
            value={password}
            onChange={(e) => setPassword(e.target.value)}
          />
        </div>
        <div className={styles.formGroup}>
          <label htmlFor={"phone"}>Phone Number (Optional):</label>
          <input
            name={"phone"}
            id={"phone"}
            type="text"
            value={phone}
            onChange={(e) => setPhone(e.target.value)}
          />
        </div>

        <button type="submit">
          {isSigningUp ? <BouncingDotsLoader /> : <span>Sign Up</span>}
        </button>
        {error && <div className={`${styles.errorMessage} error`}>{error}</div>}
      </form>
    </div>
  );
};

export default SignupForm;
