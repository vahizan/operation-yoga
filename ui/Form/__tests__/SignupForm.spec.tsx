import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import axios from "axios";
import SignupForm from "../SignupForm";
import { useRouter } from "next/navigation";

jest.mock("axios");
jest.mock("next/navigation", () => {
  return {
    ...jest.requireActual("next/router"),
    useRouter: jest.fn(),
  };
});

describe("SignupForm", () => {
  const pushMock = jest.fn();
  beforeEach(() => {
    (useRouter as jest.Mock).mockImplementation(() => ({
      push: pushMock,
    }));
    render(<SignupForm />);
  });

  it("renders the form", () => {
    const { container } = render(<SignupForm />);

    expect(container).toMatchSnapshot();
  });

  it("displays an error message if required fields are not filled", async () => {
    const submitButton = screen.getByRole("button", { name: "Sign Up" });
    fireEvent.click(submitButton);

    await waitFor(() => {
      const errorElement = screen.getByText(
        "Please fill in all required fields."
      );
      expect(errorElement).toBeInTheDocument();
    });
  });

  it("submits the form and redirects to the login page on successful signup", async () => {
    const mockResponse = { data: {} };
    (axios.post as jest.Mock).mockResolvedValueOnce(mockResponse);

    const submitButton = screen.getByRole("button", { name: "Sign Up" });

    fireEvent.change(screen.getByLabelText("Name:"), {
      target: { value: "John Doe" },
    });
    fireEvent.change(screen.getByLabelText("Email:"), {
      target: { value: "john@example.com" },
    });
    fireEvent.change(screen.getByLabelText("Password:"), {
      target: { value: "password123" },
    });

    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(axios.post).toHaveBeenCalledWith(
        "/api/auth/signup",
        JSON.stringify({
          name: "John Doe",
          email: "john@example.com",
          password: "password123",
          phone: "",
        }),
        {
          headers: {
            "Content-Type": "application/json",
          },
        }
      );
      expect(
        screen.queryByText("An error occurred. Please try again.")
      ).toBeNull();
      expect(
        screen.queryByText("Please fill in all required fields.")
      ).toBeNull();
      expect(screen.queryByText("Success message")).toBeNull();
      expect(screen.queryByText("Login page")).toBeNull();
    });
  });

  it("displays an error message if signup fails", async () => {
    const errorMessage = "Signup failed. Please try again.";
    const mockResponse = { data: { message: errorMessage } };
    (axios.post as jest.Mock).mockRejectedValueOnce(mockResponse);

    const submitButton = screen.getByRole("button", { name: "Sign Up" });

    fireEvent.change(screen.getByLabelText("Name:"), {
      target: { value: "John Doe" },
    });
    fireEvent.change(screen.getByLabelText("Email:"), {
      target: { value: "john@example.com" },
    });
    fireEvent.change(screen.getByLabelText("Password:"), {
      target: { value: "password123" },
    });

    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(
        screen.getByText("An error occurred. Please try again.")
      ).toBeInTheDocument();
      expect(
        screen.queryByText("Please fill in all required fields.")
      ).toBeNull();
      expect(screen.queryByText("Success message")).toBeNull();
      expect(screen.queryByText("Login page")).toBeNull();
    });
  });
});
