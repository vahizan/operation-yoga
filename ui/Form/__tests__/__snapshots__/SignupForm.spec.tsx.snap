// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupForm renders the form 1`] = `
<div>
  <div
    class="signupFormContainer"
  >
    <form>
      <div
        class="formGroup"
      >
        <label
          for="name"
        >
          Name:
        </label>
        <input
          id="name"
          name="name"
          type="text"
          value=""
        />
      </div>
      <div
        class="formGroup"
      >
        <label
          for="email"
        >
          Email:
        </label>
        <input
          id="email"
          name="email"
          type="email"
          value=""
        />
      </div>
      <div
        class="formGroup"
      >
        <label
          for="password"
        >
          Password:
        </label>
        <input
          id="password"
          name="password"
          type="password"
          value=""
        />
      </div>
      <div
        class="formGroup"
      >
        <label
          for="phone"
        >
          Phone Number (Optional):
        </label>
        <input
          id="phone"
          name="phone"
          type="text"
          value=""
        />
      </div>
      <button
        type="submit"
      >
        <span>
          Sign Up
        </span>
      </button>
    </form>
  </div>
</div>
`;
